packages/dapp-svelte-wallet/api/src/lib-wallet.js

```diff
@@ -526,26 +537,25 @@ export function makeWallet({
             })
             .map((payoutP, payoutIndex) => {
               const keyword = payoutIndexToKeyword[payoutIndex];
-              const purse = purseKeywordRecord[keyword];
-              if (purse && payoutP) {
+              if (payoutP) {
+                // We try to find a purse for this keyword, but even if we don't,
+                // we still make it a normal incoming payment.
+                const purseOrUndefined = purseKeywordRecord[keyword];
+
                 // eslint-disable-next-line no-use-before-define
-                return addPayment(payoutP, purse);
+                return addPayment(payoutP, purseOrUndefined);
               }
               return undefined;
             }),
         );
-      })
-      .finally(() =>
-        // Try reclaiming any payments that were left on the table.
-        keywords.forEach((keyword, i) => {
-          const purse = purseKeywordRecord[keyword];
-          if (purse && payments[i]) {
-            // eslint-disable-next-line no-use-before-define
-            addPayment(payments[i], purse);
-          }
-        }),
-      );
+      });
+
+    // Regardless of the status of the offer, we try to clean up any of our
+    // unclaimed payments.
+    depositedP.finally(claimMyLeftovers);
```

Comment:
The `keywords` are just the keys from `proposal.give`. The payout may include keywords that are not included in the `keywords` array. 

Thus, this code can still be dropping payments on the floor. I think we need to change lines 530 to 555 to not use any concept of the `keywords` array, and instead iterate over all the keys and values in the payout object.

Reply:
I've done this now, and I agree the code is clearer.

But I think maybe you misunderstood what I intended `claimMyLeftovers` for, so I've renamed it to `tryReclaimingWithdrawnPayments`.  This is because I identified two orthogonal forms of "dropped payments":

1. Zoe returns unexpected things via the `payoutObj`.  Following your advice, this path should now be clearer.
2. A payment we withdrew from one of our purses has somehow not been claimed by any other part of the system.  That's what `tryReclaimingWithdrawnPayments` is for, so that we protect the user's assets against system misbehaviour by reclaiming them on error when trying to withdraw them all, or also after the complete Zoe offer pathway (`depositedP`) settles.

Reply:
Ah yes, I misunderstood that the reclaiming after the payouts are deposited was only about the `give` payments, and not about the payouts. 

However, am I right that there was a bug prior to your PR (which you just fixed) in how the payouts were being deposited? The old code will drop any payment that doesn't have an associated keyword in `purseKeywordRecord`, which is only the `give` keywords and their purses: 

```js
      .then(payoutObj => {
        const payoutIndexToKeyword = [];
        return Promise.all(
          Object.entries(payoutObj)
            .map(([keyword, payoutP], i) => {
              payoutIndexToKeyword[i] = keyword;
              return payoutP;
            })
            .map((payoutP, payoutIndex) => {
              const keyword = payoutIndexToKeyword[payoutIndex];
              const purse = purseKeywordRecord[keyword];
              if (purse && payoutP) { // if purse is undefined, does nothing with the payout payment
                // eslint-disable-next-line no-use-before-define
                return addPayment(payoutP, purse);
              }
              return undefined;
            }),
        );
      })
```
So I think it's not just that what I suggested is clearer. I think it fixes a bug introduced prior to your PR.

Reply:
Ah, but you had changed `purse && payoutP` to just `payoutP` which already fixed it. Nice! That was a big bug.

Reply:
Sorry, I realized I was wrong about the purseKeywordRecord being the `give` only. Looks like 762 and 763 populate it with the `want` and `give`.

Reply:
Ok, you can disregard this whole thread. 😆  Resolving!